pkg/shell/shell.go

```diff
@@ -20,6 +20,11 @@ var logger = util.GetLogger("[shell] ")
 // Program is the shell subprogram.
 var Program prog.Program = program{}
 
+// This is set to true by shell unit tests so that a panic doesn't appear to
+// result in a successful test due to the default behavior of launching a
+// rescue shell.
```

Comment:
The first sentence of a godoc should always start with the symbol being described, and describe what it is or what it does in general, even if there is only one use case.

In this case, the godoc should start with something like:

```go
// PanicIsFatal controls whether a panic in the shell is considered fatal...
```

Also, this flag is kind of backward - panic being fatal is the default in Go. So instead of a `PanicIsFatal` flag, it's better to introduce a `RescuePanic` flag that defaults `true`, set it to `false` in tests, and make the `defer rescue` call conditional:

```go
if RescuePanic {
    defer rescue()
}
```

Reply:
I'll eventually learn to write Go doc strings correctly :-) I'll fix that wording.

Regarding the var name and how it is used.... My solution doesn't alter the main path and provides a way, in theory, to test that panics are handled correctly; i.e., by running the rescue function modulo not launching a rescue shell. It also opens the possibility of making the rescue shell behavior optional in normal use via a CLI flag or some other mechanism (e.g., a builtin var) while still allowing a lot of Elvish specific context being output when a panic occurs. For example, the current statement being executed, open files, etc. That might prove useful although I'll readily admit I can't provide a real world use case as I write this.

Reply:
I've pushed a change that improves the doc string. Also, looking at how `defer rescue()` is used it seems to me the use in *pkg/shell/script.go* is problematic. At least with respect to the default behavior of launching a rescue shell. That should only occur for interactive shells -- not when executing arbitrary scripts. Under what circumstances does it make sense to launch */bin/sh* as a rescue shell if something like `elvish -c 'put WTF'` results in a panic? I can't help but wonder if that path shouldn't also set `PanicIsFatal` to true. It could, of course, simply not `defer rescue()` but that would eliminate the possibility of providing more context than the default Go panic output.

Reply:
I ran a quick and dirty test by adding a `panic("WTF")` to function `evalInTTY` and ran `./elvish -c 'echo yes'`. Sure enough, it showed me the panic info and also left me in a rescue shell. It's not clear why the `Script()` path is calling `evalInTTY()` but I can't see how dropping into a rescue shell as a result of a panic is correct for that case. I added `PanicIsFatal = true` as the first statement of function `Script` and observed the expected behavior; i.e., elvish specific panic output while the shell still exits with a non-zero status rather than launching a rescue shell. So I'm going to add that to this P.R.

Reply:
Having slept on this I've decided to rename the bool. Primarily because panics are always fatal so whether `PanicIsFatal` is true or false doesn't change the fact the elvish process is terminated when a panic occurs. However, I still feel that it's better to keep the `defer rescue()` and simply predicate the launching of an actual rescue shell on whether the shell is interactive so additional context about the panic can be output (at least in theory).